### PR TITLE
Prometheus2

### PR DESCRIPTION
**Pull request by kxingit**

### CODE REVIEW

File: hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        handler.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
+        register();
+    }
+
+    /**
+     * register prometheus jmx collector.
+     */
+    private static void register() {
+        try {
+            new JmxCollector("").register();
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java]: Just curious if passing empty yaml string is a normal usage pattern?

File: hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            DefaultExports.initialize();
+            log.info("Prometheus JmxCollector registered.");
+        } catch (Exception e) {
+            log.warn("Prometheus collector not registerd: ", e);
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java]: Why not log.error? Is prometheus dependency optional?

File: hydra-main/src/main/java/com/addthis/hydra/minion/Minion.java [hunk 3]
```diff
@@ -259,8 +262,12 @@ private Minion(@JsonProperty("dataDir") File rootDir,
         }
         activeTaskKeys = new HashSet<>();
         jetty = new Server(webPort);
+
+        PrometheusServletCreator.create(handler);
+        handler.start();
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/minion/Minion.java]: Is it normal to start and stop handlers outside of Jetty? I don't see pattern this in any other code changed in this PR.

File: hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java (hunk 1 above)
Comment (jhorwit2) [reply hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java]: This should support a configuration file so you can map the horrible default dropwizard names to proper prometheus style metrics.

**Comment by jhorwit2** [comment]
Can you please squash the commits and provide a better commit message? Squashing the initial commits will help us be able to review the changes made.

File: hydra-main/src/main/java/com/addthis/hydra/job/web/SpawnService.java [hunk 4]
```diff
@@ -189,9 +190,10 @@ public void start() throws Exception {
         ServletHolder sh = new ServletHolder(servletContainer);
 
         handler.addFilter(GzipFilter.class, "/*", EnumSet.of(DispatcherType.REQUEST));
-        handler.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
+        handler.addServlet(new ServletHolder(new MetricsServlet()), "/yammer_metrics");
```
[the remark this answers is not in the document]
Comment (kxingit) [reply hydra-main/src/main/java/com/addthis/hydra/job/web/SpawnService.java]: yammer metrics servlets deleted

File: hydra-main/src/main/java/com/addthis/hydra/minion/Minion.java (hunk 3 above)
Comment (kxingit) [reply hydra-main/src/main/java/com/addthis/hydra/minion/Minion.java]: handler start moved to inside of jetty.

File: hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java (hunk 1 above)
Comment (kxingit) [reply hydra-main/src/main/java/com/addthis/hydra/util/PrometheusServletCreator.java]: Added support for configuration files.

File: hydra-main/src/main/java/com/addthis/hydra/minion/Minion.java [hunk 5]
```diff
@@ -262,16 +261,19 @@ private Minion(@JsonProperty("dataDir") File rootDir,
         }
         activeTaskKeys = new HashSet<>();
         jetty = new Server(webPort);
+
+        PrometheusServletCreator.create(handler);
         jetty.setHandler(minionHandler);
         jetty.start();
+        handler.start();
```
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/minion/Minion.java]: Add comments on why handler.start/stop are required. Also I didn't mean move the code, please put it back where you think it should go (above)

**Pull request merged**